src/box/wal.c

```diff
@@ -374,6 +376,7 @@ tx_complete_batch(struct cmsg *msg)
 	/* Update the tx vclock to the latest written by wal. */
 	vclock_copy(&replicaset.vclock, &batch->vclock);
 	tx_schedule_queue(&batch->commit);
+	trigger_run(&wal_on_batch_write, NULL);
```

Comment:
Can somehow `wal_set_watcher()` be reused? Like relays do. The problem right now is that we add this slowdown for the common path, for all WAL writes, even for non-linearizable txns.

Reply:
I was going to write that it can't, but it actually can.
You see, the problem is that watcher is notified first and only then `tx_complete_batch` message is sent. So by the time watcher is notified in tx, fresh vclock is not yet delivered there. So I can't check if target vclock is reached or not in watcher callback. I have to wake up all the waiter fibers, because by the time they are woken up vclock will be already delivered.

TBH this looks rather complex IMO. Also do we really care about a performance impact of `trigger_run` for empty trigger list when no linearizable transactions are used? There's a walk over list of finished transactions anyway.

Some pseudocode of how I see it. I didn't try to compile it, this is just to illustrate. Let me know if you like it or not.
<details><summary>example</summary>

```c

/** A vclock watcher structure. */
struct box_vclock_watcher {
	/** The base wal watcher. */
	struct wal_watcher watcher;
	/** How many fibers are subscribed to vclock updates. */
	int waiter_count;
	/** Fibers waiting for some vclock. */
	struct rlist waiters;
};

/**
 * A single watcher for all vclock waiters initialized and destroyed on demand.
 */
struct box_vclock_watcher box_vclock_watcher;

struct vclock_waiter {
	/** The fiber waiting for vclock. */
	struct fiber *fiber;
	/** A link in all registered waiters list. */
	struct rlist link;
};

/** This is called in tx_prio, just like tx_complete_batch.
 * Since we don't yield, it's fine that this callback is fired before the
 * tx_complete_batch. By the time waiters wake up they already see the new
 * vclock.
 */
static void
box_vclock_watcher_cb(struct wal_watcher *watcher, unsigned events)
{
	(void) events;
	assert(watcher == &box_vclock_watcher.watcher);
	struct vclock_waiter *waiter;
	/**
	 * Wake up everyone. Vclock is not delivered yet, it's delivered by
	 * next cbus message. But it'll be known by the time fibers wake up.
	 */
	rlist_foreach_entry(waiter, box_vclock_watcher.waiters, link) {
		fiber_wakeup(waiter->fiber);
	}
}

void
box_add_vclock_waiter(struct vclock_waiter *waiter)
{
	if (box_vclock_watcher.waiter_count++ == 0) {
		//XXX: some difficulties with watcher initialization left out.
		// It does cbus_pair inside, we don't need that for tx <-> wal.
		// They're already connected.
		wal_set_watcher(&box_vclock_watcher.watcher, "wal",
				box_vclock_watcher_cb, NULL);
	}
	rlist_add_tail_entry(&box_vclock_watcher.waiters, waiter, link);
	return;
}

void
box_remove_vclock_waiter(struct vclock_waiter *waiter)
{
	if (--box_vclock_watcher.waiter_count == 0) {
		wal_clear_watcher(&box_vclock_watcher.watcher, NULL);
	}
	rlist_del_entry(waiter, link);
}

static int
box_wait_vclock(const struct vclock *vclock, double deadline)
{
	struct vclock_waiter waiter = {
		.fiber = fiber(),
		.link = RLIST_LINK_INITIALIZER,
	};
	box_add_vclock_waiter(&waiter);
	bool done = false;
	do {
		if (fiber_yield_deadline(deadline))
			break;
		done = vclock_compare_ignore0(&replicaset.vclock, vclock) <= 0;
	} while (!done);

	box_remove_vclock_waiter(&waiter);

	if (!done) {
		diag_set(ClientError, ER_TIMEOUT);
		return -1;
	}
	return 0;
}

```
</details>

Reply:
> You see, the problem is that watcher is notified first and only then tx_complete_batch message is sent. So by the time watcher is notified in tx, fresh vclock is not yet delivered there. So I can't check if target vclock is reached or not in watcher callback.

Upon receipt of WAL notification you can do `wal_sync()` in `box_wait_vclock()` to ensure the vclock is up to date, and then read it.

As for having a global watcher with all waiters - AFAIU the problem is only that watcher attachment is expensive due to cbus_pair/unpair things, right? Maybe it could be simplified to be able attach without pairing the current thread with WAL thread? Then you could have "small" single-fiber watchers for each such txn.

I am not insisting on it. The current version also works. As for performance - after the current version surely it wouldn't get better. Currently `trigger_run()` even for an empty list will try to run 2 loops. It is also not too good for the branch predictor because these loops are not always empty or always full. It might affect perf under insane load like say 1kk txns per second. Nonetheless, I am fine with that. Up to you if you want to try the WAL watchers way.

Reply:
> Upon receipt of WAL notification you can do wal_sync() in box_wait_vclock() to ensure the vclock is up to date, and then read it.

I'll have to do it in `box_wait_vclock`, i.e. in the waiting fiber, not in the watcher callback. Because watcher callback is fired by sched fiber. Besides, I already know the new vclock in the waiting fiber. Because by the time it's woken up the fresh vclock is delivered.

The problem I wanted to point out was that the callback doesn't know if the target vclock is reached or not, and has to wake every waiter on every batch return from wal. Which is also suboptimal. We could as well make all waiters sleep for 0.01 seconds or something in a busy loop.

Now, as for small per-fiber watchers, I don't see how we can avoid doing a cbus_call to WAL to attach the watcher.
If a watcher is attached per each new linearizable request, this will clog the WAL thread queue.
Because wathcer attaches are not batched into a single cbus message like transactions.

Long story short, watcher is a good idea only if it knows which vclock was reached in its callback. Our watcher doesn't know that, so has to wake up all waiters (even with your idea about wal_sync). In this case I think that doing a `sleep(0.01)` loop without any watchers is better. Because it's much simpler, and it results in almost the same sleeps and wake-ups for waiting fibers.

Although we could try to make watchers deliver written vclock to tx thread. Do you see any problems with that?

If the watcher knows the vclock, it will wake only the necessary waiters. We will have a single watcher for all waiters, and there will be no penalty for simple transactions when no linearizable transactions are in flight.

Reply:
> If the watcher knows the vclock, it will wake only the necessary waiters. We will have a single watcher for all waiters, and there will be no penalty for simple transactions when no linearizable transactions are in flight.

I can try to make `tx_complete_batch` be sent earlier than wal_watcher_notify. Then watcher callback will know the actual vclock by the time it's scheduled.

Reply:
> I will try to make tx_complete_batch be sent earlier than wal_watcher_notify. Then watcher callback will know the actual vclock by the time it's scheduled.

Here's how it looks like (not pushed here yet, because I'm still trying to rework the watcher to use existing pipes)

<details><summary>diff</summary>

```diff
diff --git a/src/box/wal.c b/src/box/wal.c
index dae8e5f1d..af7a6ae29 100644
--- a/src/box/wal.c
+++ b/src/box/wal.c
@@ -219,7 +219,10 @@ static void
 tx_complete_batch(struct cmsg *msg);
 
 static struct cmsg_hop wal_request_route[] = {
-	{wal_write_to_disk, &wal_writer_singleton.tx_prio_pipe},
+	{wal_write_to_disk, NULL},
+};
+
+static struct cmsg_hop wal_response_route[] ={
 	{tx_complete_batch, NULL},
 };
 
@@ -236,7 +239,8 @@ wal_msg_create(struct wal_msg *batch)
 static struct wal_msg *
 wal_msg(struct cmsg *msg)
 {
-	return msg->route == wal_request_route ? (struct wal_msg *) msg : NULL;
+	return msg->route == wal_request_route ||
+	       msg->route == wal_response_route ? (struct wal_msg *) msg : NULL;
 }
 
 /** Write a request to a log in a single transaction. */
@@ -1174,8 +1178,15 @@ done:
 		assert(err_code == JOURNAL_ENTRY_ERR_UNKNOWN);
 	}
 	fiber_gc();
+	cmsg_init(msg, wal_response_route);
+	ERRINJ_NOT_SET(ERRINJ_RELAY_FASTER_THAN_TX, {
+		cpipe_push(&writer->tx_prio_pipe, msg);
+	});
 	wal_notify_watchers(writer, WAL_EVENT_WRITE);
-	ERROR_INJECT_SLEEP(ERRINJ_RELAY_FASTER_THAN_TX);
+	ERROR_INJECT(ERRINJ_RELAY_FASTER_THAN_TX, {
+		ERROR_INJECT_SLEEP(ERRINJ_RELAY_FASTER_THAN_TX);
+		cpipe_push(&writer->tx_prio_pipe, msg);
+	});
 }
 
 /** WAL writer main loop.  */
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index f015c9354..221eae5f4 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -202,6 +202,7 @@ void errinj_set_with_environment_vars(void);
 #  define ERROR_INJECT_WHILE(ID, CODE)
 #  define errinj(ID, TYPE) ((struct errinj *) NULL)
 #  define ERROR_INJECT_COUNTDOWN(ID, CODE)
+#  define ERRINJ_NOT_SET(ID, CODE) CODE
 #else
 #  /* Returns the error injection by id */
 #  define errinj(ID, TYPE) \
@@ -233,6 +234,12 @@ void errinj_set_with_environment_vars(void);
 			CODE;						\
 		}							\
 	} while (0)
+#  define ERRINJ_NOT_SET(ID, CODE)			\
+	do {						\
+		if (!errinj(ID, ERRINJ_BOOL)->bparam) {	\
+			CODE;				\
+		}					\
+	} while (0)
 #endif
 
 #define ERROR_INJECT_RETURN(ID) ERROR_INJECT(ID, return -1)

```
</details>

Reply:
As I've written in the chat, `wal_watcher` rework turned out quite badly.
`wal_watcher` owns `cpipe`s to WAL and back to watcher. TX thread already has a pipe to and from WAL, and using separate pipes for the watcher is not an option: watcher notification has to arrive via the same pipe as tx batch return.
Otherwise watcher notification and tx batch return are unordered, which makes the watcher pointless. It might be notified before vclock is updated.

Leaving watcher rework aside, the patch about making the watcher always notified **after** tx batch return is still quite ugly (see in the comment above).

Now about the watcher rework: I have to introduce some `light_watcher`, which doesn't own cpipes. This results in either a lot of code duplication or ugly attempts to reuse code for regular and light watchers.

Long story short, I propose to go back to fiber sleep loop. No on_batch_return triggers, no watchers.
This is almost busy waiting, but it's much simpler, and it has no performance impact on "normal" transactions.

Reply:
After some more discussions in chat decided to leave the current approach with on_batch_write triggers.

Reply:
Thanks for trying.